### PR TITLE
FE-3305 Handle Node HTTP2 race condition on GOAWAY

### PR DESCRIPTION
**Pull request by cleve-fauna**
Ticket(s): FE-3305

## Problem
- TLDR; In Node, there is a race condition between handling GOAWAY and submitting requests - that can cause clients that safely checked handle go away to submit requests fter a GOAWAWY was recieved anyway.
- A short idle timeout can trigger this race condition.
- We had several tests not closing clients, leaving open handles (items still waiting for events on test exit) - which could mask bugs on close.
- technical explanation: node HTTP2 request gets put on event queue before it is actually executed. In the interim, a GOAWAY can come and cause the request to fail with a GOAWAY.
## Solution
- retry on GOAWAY up to three times. I did not put a sleep in here as it should be pretty snappy. We just need to get in line properly in the event queue.
- Increase the default idle timeout.
- close all clients in tests
## Result
- less go away errors
## Testing
Used the driver sandbox over a VPN from Buenos Aires.

When I run this code in it:

```javascript
import { Client, fql, } from "fauna";

const client = new Client({ format: "decorated" });

try {
  console.log(await client.query(fql`"Hello"`));
  client.close();
} catch (e) {
  console.log(e);
  console.log(e.summary);
  console.log("abort data " + e.abort);
}
```

with this commit I get this output:

```
~/workplace/drivers-docker-sandbox/javascript (main*) » node index.js                                                                                                                                       cleve@Cleves-MBP
{
  data: '"Hello"',
  summary: '',
  txn_ts: 1683846299689199,
  stats: {
    compute_ops: 1,
    read_ops: 0,
    write_ops: 0,
    query_time_ms: 2,
    contention_retries: 0,
    storage_bytes_read: 0,
    storage_bytes_write: 0
  }
}
```

With 0.7.0 I get this output:

```
~/workplace/drivers-docker-sandbox/javascript (main*) » node index.js                                                                                                                                       cleve@Cleves-MBP
NetworkError: The network connection encountered a problem.
    at _NodeHTTP2Client.request (/Users/cleve/workplace/drivers-docker-sandbox/javascript/node_modules/fauna/dist/node/index.js:868:13)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async #query (/Users/cleve/workplace/drivers-docker-sandbox/javascript/node_modules/fauna/dist/node/index.js:1099:29)
    at async file:///Users/cleve/workplace/drivers-docker-sandbox/javascript/index.js:6:15 {
  [cause]: Error [ERR_HTTP2_GOAWAY_SESSION]: New streams cannot be created after receiving a GOAWAY
      at new NodeError (node:internal/errors:399:5)
      at ClientHttp2Stream.requestOnConnect (node:internal/http2/core:705:17)
      at node:internal/http2/core:1830:68
      at Array.forEach (<anonymous>)
      at ClientHttp2Session.<anonymous> (node:internal/http2/core:1830:11)
      at Object.onceWrapper (node:events:628:26)
      at ClientHttp2Session.emit (node:events:525:35)
      at emit (node:internal/http2/core:330:3)
      at process.processTicksAndRejections (node:internal/process/task_queues:84:21) {
    code: 'ERR_HTTP2_GOAWAY_SESSION'
  }
}


```

----
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/http-client/node-http2-client.ts [hunk 1]
```diff
@@ -64,78 +64,41 @@ export class NodeHTTP2Client implements HTTPClient {
...
+    do {
+      try {
+        return await this.#doRequest(req);
+      } catch (error: any) {
+        // see https://github.com/nodejs/node/pull/42190/files
+        // and https://github.com/nodejs/help/issues/2105
+        //
+        // TLDR; In Node, there is a race condition between handling
+        // GOAWAY and submitting requests - that can cause
+        // clients that safely checked handle go away to submit
```
Comment (cleve-fauna) [inline src/http-client/node-http2-client.ts:79]:
```suggestion
        // clients that safely handle go away to submit
```

File: src/client.ts [hunk 2]
```diff
@@ -39,7 +39,7 @@ export const DEFAULT_CLIENT_CONFIG: Omit<ClientConfiguration, "secret"> = {
   client_timeout_buffer_ms: 5000,
   endpoint: endpoints.default,
   format: "tagged",
-  http2_session_idle_ms: 500,
+  http2_session_idle_ms: 5000,
```
Comment (CLBray) [inline src/client.ts:42]: Are these idle timeout changes related to GOWAY?

File: src/http-client/node-http2-client.ts [hunk 3]
```diff
@@ -64,78 +64,41 @@ export class NodeHTTP2Client implements HTTPClient {
...
+      try {
+        return await this.#doRequest(req);
+      } catch (error: any) {
+        // see https://github.com/nodejs/node/pull/42190/files
+        // and https://github.com/nodejs/help/issues/2105
+        //
+        // TLDR; In Node, there is a race condition between handling
+        // GOAWAY and submitting requests - that can cause
+        // clients that safely handle go away to submit
+        // requests fter a GOAWAWY was recieved anyway.
```
Comment (cynicaljoy) [inline src/http-client/node-http2-client.ts:80]:
```suggestion
        // requests after a GOAWAY was received anyway.
```

File: src/client.ts (hunk 2 above)
Comment (cleve-fauna) [reply src/client.ts:42]: Yes I found the short idle timeout could trigger the race condition.

**Review by pnwpedro — APPROVED** [review]
